torch/nn/modules/loss.py

```diff
@@ -1121,6 +1125,93 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
                                ignore_index=self.ignore_index, reduction=self.reduction)
 
 
+class CrossEntropyLossWithSoftLabels(_Loss):
+    r"""This criterion combines :class:`~torch.nn.LogSoftmax` and :class:`~torch.nn.KLDivLoss` in
+    one single class. It is useful when training a classification problem with `C` classes when
+    "soft labels" are required, such as when utilizing label smoothing or blended labels.
+
+    The ``input`` is expected to contain raw, unnormalized scores for each class and the ``target``
+    should match the `input` shape and contain the target distribution over the classes.
+
+    With input :math:`x` and target :math:`y`, the unreduced (i.e. with ``reduction`` set
+    to ``none``) loss :math:`L` can be described as:
+
+    .. math::
+        L = \{l_1,\dots,l_N\}, \quad
+        l_n = y_n \cdot \left( \log y_n - \log\left(\frac{\exp(x_n)}{\sum_{i} \exp(x_i)} \right) \right)
+
+    where the index :math:`N` spans all dimensions of ``input`` and :math:`L` has the same shape as
+    ``input`` and ``target``. If :attr:`reduction` is not ``'none'`` (default ``'mean'``), then the
+    scalar loss output :math:`\ell` is defined as:
+
+    .. math::
+        \ell(x, y) = \begin{cases}
+            \operatorname{mean}(L), & \text{if reduction} = \text{`mean';} \\
+            \operatorname{sum}(L),  & \text{if reduction} = \text{`sum'.}
+        \end{cases}
+
+    .. note::
+        This criterion differs from :class:`~torch.nn.CrossEntropyLoss` in that it allows for
+        "soft labels", whereas :class:`~torch.nn.CrossEntropyLoss` requires a single class index label
+        per minibatch item. If you only need a single class label, you should use
+        :class:`~torch.nn.CrossEntropyLoss` instead, as it performs better at the cost of flexibility.
+
+    .. note::
+        Using one-hot encoded labels with this criterion is equivalent to using
+        :class:`~torch.nn.CrossEntropyLoss` with the corresponding class index labels, with one
+        caveat: when ``reduction='mean'``, the loss values returned by this criterion are a constant factor
+        of `C` (i.e. num classes) smaller than those returned by :class:`~torch.nn.CrossEntropyLoss`. See the
+        example section below for details on this.
+
+    Args:
+        reduction (string, optional): Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will
+            be applied, ``'mean'``: the mean of the output is taken,
+            ``'sum'``: the output will be summed. Default: ``'mean'``
+
+    Shape:
+        - Input: :math:`(*, C)` where :math:`*` represents any number of dimensions (including none) and
+          `C = number of classes`
+        - Target: :math:`(*, C)`, same shape as the input
+        - Output: If :attr:`reduction` is ``'none'``, then :math:`(*, C)`, otherwise scalar
```

Comment:
PyTorch CrossEntropyLoss accepts input of shape (N, C, *) and target of shape (N, *) where * is any number of dimensions. Do the shapes here make CrossEntropyLoss and CrossEntropyLossWithSoftLabels inconsistent?

Reply:
In TF images are NHWC by default (instead of NCHW in PyTorch) so that's why they don't have this problem...

Reply:
You're right that they're inconsistent, but there's a few things to consider that made this choice hard:
* As you pointed out below, target shape == input shape for `CrossEntropyLossWithSoftLabels`, so that part can't be consistent with `CrossEntropyLoss`
* We don't want to unnecessarily bake in the need for `N`, since that will have to be undone for the no-batch-dim support work. To be consistent with `CrossEntropyLoss`, this would mean supporting shapes `(C)`, `(N, C)`, `(N, C, d1)`, `(N, C, d1, d2)`, etc. (i.e. sometimes `C` is in at dim 0 and sometimes it's at dim 1). I find this to be pretty confusing and honestly don't understand why the original CE did it this way, when `d1`, `d2`, etc. are semantically just more batch dimensions.

An interesting thing to consider could be supporting a configurable location for the `C` dim (i.e. the dim that log-softmax is taken over). But we couldn't easily make the default match CE's behavior, since the location of the `C` dim changes as the number of dims changes (location is 0 or -1 for shape `(C)`, 1 or -1 for shape `(N, C)`, and 1 for shapes `(N, C, d1, ...)`).

Also, admittedly confusingly, the `C` in `NHWC` / `NCHW` refers to "channels" (e.g. red, green, blue) while the `C` here refers to "number of classes".